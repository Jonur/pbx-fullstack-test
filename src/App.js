import React, { Component } from 'react';

class App extends Component {
  componentDidMount() {
    fetch('/api/products')
      .then((res) => res.json())
      .then(data => this.setState({
        listings: data,
      }));
  }


  render() {
    return (<main data-test="product-listings">Listing page</main>);
  }
}

export default App;
