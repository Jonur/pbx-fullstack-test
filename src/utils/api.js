/**
 * @function extractJSON
 * @description Abstracts the JSON parsing from the ResponseStream to a separate function
 * to simplify unit testing and keep the code DRY for future API additions
 * @param {Object} response
 */
const extractJSON = response => response.json();

/**
 * @function getProducts
 * @description Returns local server's product listings
 */
export const getProducts = () => fetch('/api/products')
  .then(extractJSON)
  .then(data => data);