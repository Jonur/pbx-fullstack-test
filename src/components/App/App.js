import React, { Component } from 'react';
import { getProducts } from '../../utils/api';

class App extends Component {
  state = {
    listings: {},
  }

  componentDidMount() {
    getProducts()
      .then(data => this.setState({
        listings: data,
      }));
  }


  render() {
    const { listings } = this.state;

    return (
      (listings.length && <main data-test="product-listings">Listing page {listings.length}</main>)
      || <div data-test="app-loader">Loading...</div>
    );
  }
}

export default App;
