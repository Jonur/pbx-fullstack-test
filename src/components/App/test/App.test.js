import React from 'react';
import App from '../App';
import { getProducts } from '../../../utils/api';

jest.mock('../../../utils/api', () => ({
  __esModule: true,
  getProducts: jest.fn(() => Promise.resolve([
    {
      'title': 'Simple Canvas',
      'description': 'Lets your pictures speak for themselves.',
    },
    {
      'title': 'Collage Canvas',
      'description': 'Can\'t choose just one pic? Put your favourite photos on one canvas - a collage.',
    },
  ]
  )),
}));

describe('App Component', () => {
  const wrapper = mount(<App />);

  it('should render the App component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should have the Products when mounted', () => {
    expect(getProducts).toHaveBeenCalledTimes(1);
  });

  it('should render the product listings component', async () => {
    await Promise.resolve(getProducts).then(() => {
      wrapper.update();
      expect(wrapper.find('[data-test="product-listings"]')).toHaveLength(1);
    });
  });
});