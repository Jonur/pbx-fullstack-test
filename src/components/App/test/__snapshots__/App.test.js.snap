// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component should render the App component 1`] = `
<App>
  <div
    data-test="app-loader"
  >
    Loading...
  </div>
</App>
`;
